react-main/src/app/components/Game/index.js

```diff
@@ -2,53 +2,67 @@ import React, { Component } from 'react';
 
 import Board from '../Board';
 
+import { connect } from 'react-redux';
+import { selectSquare, selectHistory, getWinner } from '../../../redux/game/actions';
+
 import styles from './styles.scss';
-import calculateWinner from './utils.js';
 
 class Game extends Component {
-  state = { history: [{ squares: Array(9).fill(null) }], xIsNext: true, stepNumber: 0 };
-
-  jumpTo = move => this.setState({ stepNumber: move, xIsNext: move % 2 === 0 });
-
-  handleClick = i => {
-    const history = this.state.history;
-    const squares = history[this.state.stepNumber].squares.slice();
-    if (calculateWinner(squares) || squares[i]) {
-      return;
-    }
-
-    squares[i] = this.state.xIsNext ? 'X' : 'O';
-    this.setState({
-      history: history.concat([{ squares }]),
-      xIsNext: !this.state.xIsNext,
-      stepNumber: this.state.stepNumber + 1
-    });
+
+  handleSquareClick = i => (
+      this.props.square(i)
+  );
+
+  handleHistoryClick = step => {
+      return this.props.historySelected(step);
+  };
+
+  componentDidMount(){
+     this.props.winner()
+  };
+
+  componentWillUpdate(){
+     this.props.winner()
   };
 
   render() {
-    const history = this.state.history;
-    const squares = history[this.state.stepNumber].squares;
-    const winner = calculateWinner(squares);
-    const moves = history.map((step, move) => {
+    const moves = this.props.history.map((step, move) => {
       const desc = move ? `Go to move #${move}` : 'Go to game start';
       return (
         <li key={move}>
-          <button onClick={() => this.jumpTo(move)}>{desc}</button>
+          <button onClick={() => this.handleHistoryClick(move)}>{desc}</button>
```

Comment:
We try to avoid inline functions in event handlers, because each time this component renders, a new function will be instanced. What you should do is the following:


```
// Move/index.js

// This receives the following props: the onClick handler, the move index and the description as children
class Move extends Component {
  handleClick = () => this.props.onClick(this.props.move);

  render() {
    // Here you render what you were rendering inside the map function in Game
    const desc = move ? `Go to move #${move}` : 'Go to game start';
    return (
      <li key={move}>
        <button onClick={this.handleClick}>{children}</button>
      </li>
    )
  }
}

// Game/index.js
render() {
  const moves = this.props.history.map((step, move) => {
    <Move key={step.id} onClick={this.handleHistoryClick} />
      {desc}
    </Move>
  }
}
```

Now the instead of `onClick={() => this.someFunc()}` (which created a new function each render) you are calling a fixed reference of a function.